README.md
# video-downloader
A standalone application for downloading videos from YouTube based on the [pytube](https://pytube.io/en/latest/#) library

## Features
- Select video destination folder
- Check video information (enter) and download it with the hit of a button

## Dependencies
- pytube
- PySide6



## Contributors
- [Jean Poletto](https://github.com/poletts)


CHANGELOG.md
# [1.0.1] - 2022-07-30
## TODO
- single exe file not working: pyinstaller cannot find file src/style.qss

# [1.0.0] - 2022-07-01

## Added
- interface with YT colors
- Open button to select destination folder
- On pressing "Enter" adter paste video url it shows basic video information
- Donwload button to download video to the selected destination folder

src/ui.py
"""
User interface download of YouTube videos
"""
import sys, os

if __name__ == "__main__":
    sys.path.append(os.path.abspath('.'))

from PySide6.QtWidgets import (QApplication, QWidget, QFileDialog, QStatusBar,
                               QVBoxLayout, QHBoxLayout,                                
                               QLabel, QPushButton, QLineEdit)
from PySide6.QtCore import Signal

from src import fun


class Main(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Video Downloader")
        self.central_widget = CentralFrame(self)
        self.status_bar = StatusBar(self)
        self.setup_layout()
        self.resize(600, 400)  
        # connect signal in CentralFrame to slot in StatusBar
        self.central_widget.send_message.connect(self.status_bar.showMessage)
    
    def setup_layout(self):
        lay = QVBoxLayout()
        lay.addWidget(self.central_widget)
        lay.addWidget(self.status_bar)
        lay.setSpacing(0)
        lay.setContentsMargins(0,0,0,0)        
        self.setLayout(lay)        
        with open('src\style.qss',"r") as f:            
            custom_style_sheet = f.read()
            self.setStyleSheet(custom_style_sheet)        


class CentralFrame(QWidget):
    
    send_message = Signal(str, int)

    def __init__(self, parent= None):
        super().__init__(parent)
        self.resize(600, 300)
        # Instances
        self.default_path = 'C:/Users/install/Downloads'
        # Widgets
        self.path = QLineEdit(self.default_path)
        self.open_explorer = QPushButton('Open')
        self.video_link = QLineEdit()
        self.download_btn = QPushButton('Download')
        self.video_info = QLabel()        
        self.setup_layout()
        # Connections
        self.open_explorer.clicked.connect(self.select_folder_path)
        self.video_link.editingFinished.connect(self.display_info)
        self.download_btn.clicked.connect(self.download_fun)

    def setup_layout(self):
        folder = QHBoxLayout()
        folder.addWidget(QLabel('Destination Folder: '))
        folder.addWidget(self.path)
        folder.addWidget(self.open_explorer)

        video = QHBoxLayout()
        video.addWidget(QLabel('Enter video url: '))
        video.addWidget(self.video_link)
        video.addWidget(self.download_btn)

        lay = QVBoxLayout()
        lay.addLayout(folder)
        lay.addLayout(video)
        lay.addWidget(self.video_info)
        lay.setSpacing(0)
        lay.addStretch(1)
        self.setLayout(lay)

    def select_folder_path(self):
        """Open dialog to select folder path"""

        file_path = QFileDialog.getExistingDirectory(self, 'Please select path', self.default_path)
        if file_path == '':
            self.path.setText(self.default_path)
        else:
            self.path.setText(file_path)  
    
    def display_info(self):
        """Display video info on screen"""

        link = self.video_link.text()
        if fun.check_youtube_url(link):
            txt = fun.get_video_info(link)
            self.video_info.setText(txt)
            self.send_message.emit('URL provided is valid', 5000)         
        else:
            self.send_message.emit('Invalid url', 5000)

    def download_fun(self):

        link = self.video_link.text()
        path = self.path.text()

        if fun.check_youtube_url(link):        
            self.send_message.emit('Download started...', 5000)            
            fun.import_video(link, path)
            self.video_info.setText('')
            self.send_message.emit('Video sucessfully downloaded', 5000)
        else:
            self.send_message.emit('Invalid url', 5000)  


class StatusBar(QStatusBar):
    def __init__(self, parent=None):
        super().__init__(parent)
        # Display copyright and Version
        self.lic = QLabel(self)
        self.lic.setText("\xa9 2022 poletts | v 0.0.1 ")
        self.addPermanentWidget(self.lic)
        self.showMessage('Welcome', 3000)
    
    def update(self, text:str = None) -> None:
        if text is not None:
            self.showMessage(text, 3000)

    def clean(self):
        self.clearMessage()


if __name__ == "__main__":

    app = QApplication(sys.argv)

    main = Main()
    main.show()

    sys.exit(app.exec())


src/fun.py
"""
Collection of functions interacting with pytube library
"""

import os

from pytube import YouTube

def import_video(link: str, path: str):
    """
    Donwload video from link to path

    Parameters:
    ----------
    link:str
        url with link for youTube video
    path: str
        file path for download video
        
    Return:
    ------
    None
    """
    # Stablish connection
    yt=YouTube(link);    
    # Get the highest resolution    
    ys=yt.streams.get_highest_resolution()
    #download
    ys.download(path)    
    
def check_youtube_url(url:str) -> bool:
    """
    Check if the url is accessible by pytube.YouTube module

    Parameters:
    -----------
    url : str
        url for YouTube video

    Return:
    -------
    bool
        True if url is acessible
        False if url is not accessible
    """
    try:
        YouTube(url)
        status = True
    except:
        status = False
    return status

def get_video_info(link:str) -> str:
    """
    Retrieve video informations

    Parameters:
    -----------
    link: str
        url with link for youTube video

    Return:
    -------
    txt: str
        string with information about the video 
    """
    yt=YouTube(link)
    txt = '\nTitle: ' + yt.title + '\n' \
          'Author: ' + yt.author + '\n' \
          'Duration: ' + str(yt.length) +' seconds' + '\n' \
          'Published on ' + yt.publish_date.strftime("%B %d, %Y") + '\n' \
          'Views: %d' % yt.views

    return ''.join(txt)

if __name__ == "__main__":

    # Get current path
    path=os.path.dirname(os.path.abspath(__file__))
    
    # Select link
    link=input('Insert video link:')
    
    # Import Video
    import_video(link, path)

    print('Video sucessfully downloaded')


src/youtube_downloader.py
"""
User interface for quick download of YouTube videos
"""
import sys

from pytube import YouTube
from PySide6.QtWidgets import (QApplication, QWidget, QFileDialog, QStatusBar,
                               QVBoxLayout, QHBoxLayout,                                
                               QLabel, QPushButton, QLineEdit)
from PySide6.QtCore import Signal


class Main(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setWindowTitle("YouTube Downloader")
        self.central_widget = CentralFrame(self)
        self.status_bar = StatusBar(self)
        self.setup_layout()
        self.resize(600, 400)  
        # connect signal in CentralFrame to slot in StatusBar
        self.central_widget.send_message.connect(self.status_bar.showMessage)
        self.show()
    
    def setup_layout(self):
        lay = QVBoxLayout()
        lay.addWidget(self.central_widget)
        lay.addWidget(self.status_bar)
        lay.setSpacing(0)
        lay.setContentsMargins(0,0,0,0)        
        self.setLayout(lay)
        custom_style_sheet = """
        QWidget{background-color: white;
                padding: 2px}

        QLineEdit={
                   background-color: lightblue;
                   }
        """
        self.setStyleSheet(custom_style_sheet)


class CentralFrame(QWidget):
    
    send_message = Signal(str, int)

    def __init__(self, parent= None):
        super().__init__(parent)
        self.resize(600, 300)
        # Instances
        self.default_path = 'C:/Users/install/Downloads'
        # Widgets
        self.path = QLineEdit(self.default_path)
        self.open_explorer = QPushButton('Open')
        self.video_link = QLineEdit()
        self.download_btn = QPushButton('Download')
        self.video_info = QLabel()        
        self.setup_layout()
        # Connections
        self.open_explorer.clicked.connect(self.select_folder_path)
        self.video_link.editingFinished.connect(self.display_info)
        self.download_btn.clicked.connect(self.download_fun)

    def setup_layout(self):
        folder = QHBoxLayout()
        folder.addWidget(QLabel('Destination Folder: '))
        folder.addWidget(self.path)
        folder.addWidget(self.open_explorer)

        video = QHBoxLayout()
        video.addWidget(QLabel('Enter video url: '))
        video.addWidget(self.video_link)
        video.addWidget(self.download_btn)

        lay = QVBoxLayout()
        lay.addLayout(folder)
        lay.addLayout(video)
        lay.addWidget(self.video_info)
        lay.setSpacing(0)
        lay.addStretch(1)
        self.setLayout(lay)

    def select_folder_path(self):
        """Open dialog to select folder path"""

        file_path = QFileDialog.getExistingDirectory(self, 'Please select path', self.default_path)
        if file_path == '':
            self.path.setText(self.default_path)
        else:
            self.path.setText(file_path)  
    
    def display_info(self):
        """Display video info on screen"""

        link = self.video_link.text()        

        if check_youtube_url(link):
            yt = YouTube(link)

            txt = '\nTitle: ' + yt.title + '\n' \
                  'Author: ' + yt.author + '\n' \
                  'Duration: ' + str(yt.length) +' seconds' + '\n' \
                  'Pubished on ' + yt.publish_date.strftime("%B %d, %Y") + '\n' \
                  'Views: %d' % yt.views
            self.video_info.setText(''.join(txt))   
            self.send_message.emit('URL provided is valid', 5000)         
        else:
            self.send_message.emit('Invalid url', 5000)

    def download_fun(self):

        link = self.video_link.text()
        path = self.path.text()

        if check_youtube_url(link):
            self.send_message.emit('Download started...', 5000)            
            yt=YouTube(link)
            # Get the highest resolution    
            ys=yt.streams.get_highest_resolution()    
            #download
            ys.download(path)
            self.video_info.setText('')
            self.send_message.emit('Video sucessfully downloaded', 5000)
        else:
            self.send_message.emit('Invalid url', 5000)  


class StatusBar(QStatusBar):
    def __init__(self, parent=None):
        super().__init__(parent)
        # License
        self.lic = QLabel(self)
        self.lic.setText("(C) 2022 Jean Poletto ")
        self.addPermanentWidget(self.lic)
        # Version
        self.version = QLabel(self)
        self.version.setText(" v0.0.1 ")
        self.addPermanentWidget(self.version)        
        self.showMessage('Welcome', 3000)
    
    def update(self, text:str = None) -> None:
        if text is not None:
            self.showMessage(text, 3000)

    def clean(self):
        self.clearMessage()


def check_youtube_url(url:str) -> bool:
    """
    Check wheter an url is accessible by pytube.YouTube module

    Parameters:
    -----------
    url : str
        url for YouTube video

    Return:
    -------
    bool
        True if url is acessible
        False if url is not accessible
    """
    try:
        YouTube(url)
        status = True
    except:
        status = False
    return status


if __name__ == "__main__":

    app = QApplication(sys.argv)

    main = Main()

    sys.exit(app.exec())
